File: CompetitiveProgramming/Arrays/rotateImage.js
/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */


 var rotate = function(matrix) {
   
    transpose(matrix)
    reflect(matrix)
    
    
};

var transpose = function(matrix){
    
    const n = matrix.length
    
    for(let i = 0 ; i <n ; i++ ){
        
        for(let j = i+1 ; j <n ; j++){
            
            const temp =  matrix[i][j]
            matrix[i][j] = matrix[j][i]
            matrix[j][i] = temp
        }
            
    }
    
    
}


var reflect = function(matrix){
    
    const n = matrix.length
    
    for(let i = 0; i<n ; i++){
        
        for(let j = 0; j< n/2 ; j++){
                
            temp = matrix[i][j]
            
            matrix[i][j] = matrix[i][n-j-1]
            
            matrix[i][n-j-1] = temp

            
        }
                        
    }
}

File: CompetitiveProgramming/Arrays/rotateImage.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """

        rows = len(matrix)
        cols = len(matrix[0])

        def transpose(matrix):
            for i in range(rows):
                for j in range(i,cols):
                    temp = matrix[j][i]
                    matrix[j][i] = matrix[i][j]
                    matrix[i][j] = temp

        def reflect(matrix):
            for i in range(rows):
                for j in range(cols//2):
                    temp = matrix[i][cols-1-j]
                    matrix[i][cols-1-j] = matrix[i][j]
                    matrix[i][j] = temp 



        transpose(matrix)
        reflect(matrix)



